### PR TITLE
[JENKINS-68780] Blocked upstream projects in the queue block downstream projects 

### PR DESCRIPTION
**Pull request by Si-So**
See [JENKINS-68780](https://issues.jenkins.io/browse/JENKINS-68780).

### Proposed changelog entries

* Blocked upstream projects in the queue block downstream projects when the option "Block build when upstream project is building" is enabled for the downstream project.
* Blocked downstream projects in the queue block upstream projects when the option "Block build when downstream project is building" is enabled for the upstream project.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate. 
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@res0nance
@basil 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/6675"></a>

### CODE REVIEW

**Comment by Si-So** [comment]
This first commit only includes a test case. This test case is a translation of the ticket's text into code. 

I expect only my new test to fail.

**Comment by Si-So** [comment]
I am currently confused by the results of ci.jenkins.io. I am waiting for more runs to complete because right now I am not sure if the problems are caused by my PR or are by a problem of ci.jenkins.io itself.

**Comment by basil** [comment]
Your code is fine. The system is just unstable. :cry:

**Review by res0nance — DISMISSED** [review]

**Comment by Si-So** [comment]
@res0nance: Thanks for the approval, but I think it was premature: After adding the two latest commits an old test started to fail and it is quite likely that this due to my changes. 

This failure was probably hidden because ci.jenkins.io was unstable. 

I will look into the issue.

**Comment by Si-So** [comment]
The core of the problem is that two contradictory requirements exist. A simplified example: 

* projectA triggers projectB and projectC
* projectB triggers projectC
* projectC triggers projectB
* Both projectB and projectC have "Block build when upstream project is building" enabled

(I am not sure, if this example actually triggers the problem or if a more complex example is needed. But due to its simplicity, it makes the core of the problem easier to understand)

The current behavior of jenkins is that the circular dependency of projectB and projectC is detected and the cycle is broken and one of the projects starts to build. This is in contrast to the described behavior of  "Block build when upstream project is building". When this behavior was altered in a meaningful way for the last time, the committer also mentions this. I quote: 
>One could argue that without this change the system is functioning correctly and that previous behaviour was a bug. On the other hand, people have come to rely on the previous behaviour.

See: https://github.com/jenkinsci/jenkins/commit/4f4a64a522ec7bf31f24280827757214e6985f3d

This clashes with my requirement that the options work as described. 

Whereas the current behavior reduces the chances of deadlocks occurring I am fine with deadlocks occurring. 

I propose the following solution: 
Add a "strict mode" sub-option for the option "Block build when upstream project is building". When this sub-option is enabled my desired behavior is toggled on. I like this solution because it is backward compatible. I believe the hardest part of this solution is to explain the difference between non-strict-mode and strict-mode in the UI. 

I am a little bit unsure if my proposed solution will be merged since it has effects on the UI. 

Will my solution be merged after I have finished it?

**Comment by Si-So** [comment]
I just pushed a new commit. 
When the queue checks, if a project can be built upstream/downstream projects that are blocked themselves by upstream or downstream builds, are ignored to break deadlocks. Ignoring upstream or downstream builds is fine since getBuildingUpstream()/getBuildingDownstream() will eventually find the root cause of the blockage and will thus break the circular dependency by starting the oldest scheduled project of the cycle. This is also how jenkins master currently handles things.

**Comment by Si-So** [comment]
Ci.Jenkins.io looks good. I just pushed updated help texts.

**Comment by Si-So** [comment]
Can I support the review process in any way?

**Review by res0nance — APPROVED** [review]
LGTM

**Comment by basil** [comment]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Review by basil — APPROVED** [review]

**Comment by daniel-beck** [comment]
The labels aren't changed, so I would expect some (many) users to be surprised about the new behavior. Is this change worth it? What is the motivation for this change other than it doesn't work like the documentation says? Do we believe it should behave as documented rather than as implemented?

File: core/src/main/java/hudson/model/AbstractProject.java
Comment (jtnord) [inline core/src/main/java/hudson/model/AbstractProject.java:1126]: please ensure any javadoc is updated to match the expected contract.
the change here is that if the upstream is blocked this no longer is a potential candidate.  (the previous javadoc may also have been incorrect)

e.g.
```
    /**
     * Returns the project if any of the upstream project is either
     * building, or scheduled and not blocked by an upstream/downstream project build.
     * <p>
     * This means eventually there will be an automatic triggering of
     * the given project (provided that all builds went smoothly.)
     */
```

**Review by jtnord — COMMENTED** [review]
> The labels aren't changed, so I would expect some (many) users to be surprised about the new behaviour. Is this change worth it? What is the motivation for this change other than it doesn't work like the documentation says? Do we believe it should behave as documented rather than as implemented? Otherwise, changing the documentation would be the potentially less harmful choice.

Probably needs to be answered, however the code change looks good (pending javadoc tweaks) and if it breaks installations we could revert it in future.

Of note - if I understand this correctly you should be able to reproduce the reported issue without any external plugins blocking builds.  If a build is blocked because all nodes with a given label have no spare executors  (yes most agents would come from some plugin but the blockage is defined in Jenkins core not the agent/cloud plugin.)

And given that - I think it makes sense that up/downstreams are blocked whilst a up/downstream is waiting for an agent.

File: core/src/main/java/hudson/model/AbstractProject.java
Comment (Si-So) [reply core/src/main/java/hudson/model/AbstractProject.java:1126]: Done

File: core/src/main/java/hudson/model/AbstractProject.java
Comment (Si-So) [reply core/src/main/java/hudson/model/AbstractProject.java:1109]: Done

**Review by jtnord — APPROVED** [review]

**Comment by Si-So** [comment]
> Of note - if I understand this correctly you should be able to reproduce the reported issue without any external plugins blocking builds. If a build is blocked because all nodes with a given label have no spare executors (yes most agents would come from some plugin but the blockage is defined in Jenkins core not the agent/cloud plugin.)
> 
> And given that - I think it makes sense that up/downstreams are blocked whilst a up/downstream is waiting for an agent.

I checked the bug tracker and found multiple longstanding critical bugs that very likely occur because of the current behavior of the blocking option. These are the issues: 
* [JENKINS-22800](https://issues.jenkins.io/browse/JENKINS-22800)
* [JENKINS-5150](https://issues.jenkins.io/browse/JENKINS-5150)
* [JENKINS-5125](https://issues.jenkins.io/browse/JENKINS-5125)

Because of these issues I am strongly in favor of merging my PR.

File: test/src/test/java/hudson/model/QueueTest.java
Comment (basil) [inline test/src/test/java/hudson/model/QueueTest.java:604]: This looks like a pretty tight loop that could easily chew up a CPU core while waiting for the condition to be true. I think it would be a good idea to add a small sleep between repetitions. This would yield the CPU to the OS scheduler and loosen this tight loop.

**Review by basil — APPROVED** [review]
The motivation is clear to me; thank you for fixing this long-standing issue!

File: test/src/test/java/hudson/model/QueueTest.java
Comment (Si-So) [reply test/src/test/java/hudson/model/QueueTest.java]: Done

**Comment by Si-So** [comment]
FYI: I will be on vacation for the next two weeks. I won't respond to any change requests during this time.

**Comment by basil** [comment]
On July 13, @daniel-beck [removed the `ready-for-merge` label](https://github.com/jenkinsci/jenkins/pull/6675#event-6991403640) and [added the `needs-justification` label](https://github.com/jenkinsci/jenkins/pull/6675#event-6991405147). On July 15, @Si-So [provided a justification](https://github.com/jenkinsci/jenkins/pull/6675#issuecomment-1185474328). As of July 18, @daniel-beck has not responded to the post providing a justification, nor has he removed the `needs-justification` label.

**Comment by daniel-beck** [comment]
If others disagree, feel free to update the PR accordingly.

**Comment by basil** [comment]
Thanks Daniel! At this point all design feedback has been addressed, and all implementation feedback has also been addressed, so I think it is appropriate to restart the integration timer.

---

This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Pull request merged**
